### PR TITLE
Support automatic unsealing using AWS Secrets Manager.

### PR DESCRIPTION
**Pull request by rgooch**

### CODE REVIEW

File: cmd/keymasterd/config.go [hunk 1]
```diff
@@ -40,6 +41,8 @@ type baseConfig struct {
 	TLSCertFilename              string   `yaml:"tls_cert_filename"`
 	TLSKeyFilename               string   `yaml:"tls_key_filename"`
 	SSHCAFilename                string   `yaml:"ssh_ca_filename"`
+	UnsealPasswordSecretArn      string   `yaml:"unseal_password_secret_arn"`
```
Comment (cviecco) [inline cmd/keymasterd/config.go]: For clarity these should be in a separate section. (will make the names more readable). I would say:
type SelfUnsealing struct {
        AWSSecretName string `yaml:"aws_secret_name"`
        AWSSecertKey ....
}

And also I would use a name instead of an arn so that I could use the same config across aws regions.

File: cmd/keymasterd/config.go [hunk 2]
```diff
@@ -327,33 +333,33 @@ func loadVerifyConfigFile(configFilename string) (*RuntimeState, error) {
...
 		if !strings.HasPrefix(fileAsString, "-----BEGIN PGP MESSAGE-----") {
-			err = errors.New("Have a client CA but the CA file does NOT look like and PGP file")
-			return nil, err
+			return nil, errors.New("CA private key file does NOT look like PGP")
+		}
+		logger.Println("Starting up in sealed state")
+		if runtimeState.ClientCAPool == nil {
+			logger.Println("No client CA: manual unsealing not possible")
+		}
+		if err := runtimeState.tryAutoUnseal(); err != nil {
```
Comment (cviecco) [inline cmd/keymasterd/config.go]: Why is this blocking? I think this should be goroutine, that IF detects is on AWS will keep retrying until the system is unsealed.

File: cmd/keymasterd/unseal.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+)
+
+func (state *RuntimeState) secretInjectorHandler(w http.ResponseWriter,
```
Comment (cviecco) [inline cmd/keymasterd/unseal.go]: There are tests for the injection handler in main_test.go... make a new file unseal_test.go and move that test to that new file.

File: cmd/keymasterd/unseal.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			"Invalid Post, "+err.Error())
+		logger.Println(err)
+		return
+	}
+	w.WriteHeader(200)
+	fmt.Fprintf(w, "OK\n")
+	//fmt.Fprintf(w, "%+v\n", r.TLS)
+}
+
+func (state *RuntimeState) tryAutoUnseal() error {
```
Comment (cviecco) [inline cmd/keymasterd/unseal.go]: make this function name aws specific.

File: cmd/keymasterd/unseal.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	baseConfig := state.Config.Base
+	if baseConfig.UnsealPasswordSecretArn == "" {
+		return nil
+	}
+	secretArn, err := arn.Parse(baseConfig.UnsealPasswordSecretArn)
+	if err != nil {
+		return err
+	}
+	if baseConfig.UnsealPasswordSecretKey == "" {
```
Comment (cviecco) [inline cmd/keymasterd/unseal.go]: This should be poart of the config loading.. we dont modify configs on the consumer

File: cmd/keymasterd/unseal.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+		baseConfig.UnsealPasswordSecretKey = "UnsealPassword"
+	}
+	creds := credentials.NewCredentials(&ec2rolecreds.EC2RoleProvider{
+		Client:       ec2metadata.New(session.New(&aws.Config{})),
```
Comment (cviecco) [inline cmd/keymasterd/unseal.go]: I do this differently which I think is clearer:
func getAWSRegion() (string, error) {
        svc := ec2metadata.New(session.New())
        if !svc.Available() {
                return "", fmt.Errorf("not running on AWS or metadata not available")
        }
        return svc.Region()
}

and then:
    region, err := getAWSRegion()
        if err != nil {
                return nil, err
        }
        sess, err := session.NewSession(&aws.Config{
                Region: aws.String(region),
        })
        if err != nil {
                log.Printf("failed Sessionm err=%s", err)
                return nil, err
        }
        //Create a Secrets Manager client
        svc := secretsmanager.New(sess)
        input := &secretsmanager.GetSecretValueInput{
                SecretId:     aws.String(s.config.Unsealing.AWSSecretName),
        }

File: cmd/keymasterd/unseal.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	output, err := awsService.GetSecretValue(&input)
+	if err != nil {
+		return fmt.Errorf("error calling secretsmanager:GetSecretValue: %s",
+			err)
+	}
+	var secret []byte
+	if output.SecretString != nil {
+		secret = []byte(*output.SecretString)
+	} else {
```
Comment (cviecco) [inline cmd/keymasterd/unseal.go]: make it so that we only allow string secrets

File: cmd/keymasterd/unseal.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+		if err != nil {
+			return fmt.Errorf("base64 decode error: %s", err)
+		}
+		secret = binaryData[:length]
+	}
+	var secrets map[string]string
+	if err := json.Unmarshal(secret, &secrets); err != nil {
+		return fmt.Errorf("error unmarshaling secret: %s", err)
+	}
+	if password, ok := secrets[baseConfig.UnsealPasswordSecretKey]; !ok {
```
Comment (cviecco) [inline cmd/keymasterd/unseal.go]: this construction is hard to parse, change it to:
password, ok := secrets[baseConfig.UnsealPasswordSecretKey];
if !ok {
return fmt.Errorf(...
}
return state.unsealCA.....

File: cmd/keymasterd/config.go (hunk 1 above)
Comment (rgooch) [reply cmd/keymasterd/config.go]: Done.

File: cmd/keymasterd/config.go (hunk 2 above)
Comment (rgooch) [reply cmd/keymasterd/config.go]: I did it this way because it's really fast. However, I've changed it to a loop so that if there is a policy error or some such, Keymaster will keep trying until success.

File: cmd/keymasterd/unseal.go (hunk 3 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: Done.

File: cmd/keymasterd/unseal.go (hunk 4 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: Done.

File: cmd/keymasterd/unseal.go (hunk 5 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: It's not changing the config, only a local copy.

File: cmd/keymasterd/unseal.go (hunk 6 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: Done.

File: cmd/keymasterd/unseal.go (hunk 7 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: Done.

File: cmd/keymasterd/unseal.go (hunk 8 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: Done.

File: cmd/keymasterd/unseal.go (hunk 5 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: As discussed offline, I've split out the setting of the default.

File: cmd/keymasterd/unseal.go (hunk 6 above)
Comment (rgooch) [reply cmd/keymasterd/unseal.go]: As discussed offline, I've added a TODO.

Comment (rgooch) [reply cmd/keymasterd/unseal.go]: I've simplified this as suggested and split out the core code that deals with AWS Secrets Manager so that it can be re-used for other secrets (like the database password).
